File: ruby/functional_list.rb
module FunctionalList
  def FunctionalList.nil
    Nil.new
  end

  def FunctionalList.cons(head, tail)
    Cons.new head, tail
  end

  def FunctionalList.of(*items)
    fromArray items, 0, items.length
  end

private

  def FunctionalList.fromArray(items, start, finish)
    start == finish ? FunctionalList.nil : cons(items[start], fromArray(items, start + 1, finish))
  end
end

class Nil
  include FunctionalList

  def isEmpty
    true
  end

  def ==(other)
    other.is_a? Nil
  end

  def map(&mapping)
    self
  end
end

class Cons
  include FunctionalList

  attr_reader :head, :tail

  def initialize(head, tail)
    @head = head
    @tail = tail
  end

  def isEmpty
    false
  end

  def ==(other)
    head == other.head && tail == other.tail
  end

  def map(&mapping)
    FunctionalList.cons(mapping.call(head), tail.map(&mapping))
  end
end

File: javascript/lib/functional_list.js
(function() {
    var root = this;


    function FunctionalList() { }

    FunctionalList.nil = function() {
        return new Nil();
    };

    FunctionalList.cons = function(head, tail) {
        return new Cons(head, tail);
    };

    FunctionalList.of = function() {
        var items = Array.prototype.slice.call(arguments);
        return fromArray(items, 0, items.length);
    };

    function fromArray(items, start, finish) {
        return start === finish ? FunctionalList.nil() : FunctionalList.cons(items[start], fromArray(items, start + 1, finish));
    }


    function Nil() { }

    Nil.prototype = new FunctionalList();

    Nil.prototype.isEmpty = function() {
        return true;
    };

    Nil.prototype.map = function(mapping) {
        return this;
    };

    Nil.prototype.equals = function(other) {
        return other.isEmpty();
    };


    function Cons(head, tail) {
        this.head = function() {
            return head;
        };

        this.tail = function() {
            return tail;
        };
    }

    Cons.prototype = new FunctionalList();

    Cons.prototype.isEmpty = function() {
        return false;
    };

    Cons.prototype.map = function(mapping) {
        return FunctionalList.cons(mapping(this.head()), this.tail().map(mapping));
    };

    Cons.prototype.equals = function(other) {
        return !other.isEmpty() && this.head() == other.head() && this.tail().equals(other.tail());
    };


    if (typeof module !== 'undefined' && module.exports) {
        module.exports = FunctionalList;
    }

    root.FunctionalList = FunctionalList;
})();
